File: ReverseLinkedList.cpp
// ReverseLinkedList
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    ListNode* reverseList(ListNode* head) {
        ListNode* prev=nullptr;
        ListNode* curr=head;
        ListNode* temp;

        while(curr){
            temp=curr->next;
            curr->next=prev;
            prev=curr;
            curr=temp;
        }

        return prev;
    }

    // ListNode* reverseList(ListNode* head) {
    //     if(head==nullptr||head->next==nullptr)return head;
    //     ListNode*temp=reverseList(head->next);
    //     head->next->next=head;
    //     head->next=nullptr;
    //     return temp;
    // }
};

File: Reverse_Linked_List.kt
/**
 * Example:
 * var li = ListNode(5)
 * var v = li.`val`
 * Definition for singly-linked list.
 * class ListNode(var `val`: Int) {
 *     var next: ListNode? = null
 * }
 */
class Solution {
    fun reverseList(head: ListNode?): ListNode? {
        head ?: return null
        var node = head
        var prev: ListNode? = null
        var next: ListNode? = null
        while (node != null) {
            next = node.next
            node?.next = prev
            prev = node
            node = next
        }
        return prev
    }
}